nifi-nar-bundles/nifi-slack-bundle/nifi-slack-processors/src/main/java/org/apache/nifi/processors/slack/FetchSlack.java

```diff
@@ -0,0 +1,295 @@
+  public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+    return DESCRIPTORS;
+  }
+
+  @OnScheduled
+  public void createClient(final ProcessContext context) {
+    Header header = new BasicHeader(
+      HttpHeaders.AUTHORIZATION, "Bearer " + context.getProperty(API_TOKEN).getValue());
+    List<Header> headers = Collections.singletonList(header);
+    client = HttpClients.custom().setDefaultHeaders(headers).build();
+  }
+
+  @OnUnscheduled
+  @OnShutdown
+  public void releaseClient() {
+    try {
+      client.close();
+    } catch (IOException e) {
+      getLogger().warn("Error while closing HttpClient", e);
+    }
+    client = null;
+  }
+
+  @Override
+  public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+
+    FlowFile requestFlowFile = session.get();
+    if (requestFlowFile == null) {
+      context.yield();
+
+      return;
+    }
+
+    try {
+      List<JsonObject> filesToDownload;
+      try (InputStream requestInputStream = session.read(requestFlowFile)) {
+        filesToDownload = collectFilesToDownload(requestInputStream);
+      }
+      if (filesToDownload == null || filesToDownload.isEmpty()) {
+        session.transfer(requestFlowFile, REL_FAILURE);
+      } else {
+        downloadFiles(session, requestFlowFile, filesToDownload);
+      }
+    } catch (JsonException | IOException e) {
+      session.transfer(requestFlowFile, REL_FAILURE);
+    }
+  }
+
+  private void downloadFiles(ProcessSession session, FlowFile requestFlowFile, List<JsonObject> filesToDownload) {
+    try {
+      ArrayList<FlowFile> successfulResponses = new ArrayList<>();
+      for (JsonObject file : filesToDownload) {
+        String url = file.getString("url_private_download");
+
+        HttpUriRequest request = RequestBuilder.get().setUri(url).build();
+        HttpResponse response = client.execute(request);
+        int statusCode = response.getStatusLine().getStatusCode();
+
+        if (inStatusCodeFamily(200, statusCode)) {
+          FlowFile responseFlowFile = session.create(requestFlowFile);
+          session.write(responseFlowFile, out -> response.getEntity().writeTo(out));
+
+          Map<String, String> attributes = fillAttributes(requestFlowFile.getAttributes(), file, response);
+          responseFlowFile = session.putAllAttributes(responseFlowFile, attributes);
+
+          session.getProvenanceReporter().fetch(responseFlowFile, url);
+          successfulResponses.add(responseFlowFile);
+        } else if (inStatusCodeFamily(100, statusCode)
+          || inStatusCodeFamily(300, statusCode)
+          || inStatusCodeFamily(400, statusCode)) {
+          session.transfer(requestFlowFile, REL_NO_RETRY);
+          break;
+        } else if (inStatusCodeFamily(500, statusCode)) {
+          session.transfer(requestFlowFile, REL_RETRY);
+          break;
+        }
+      }
+
+      if (successfulResponses.size() == filesToDownload.size()) {
```

Comment:
Do we need this if statement? What if this condition doesn't match??

Reply:
In that case, NiFi may throw an Exception as the already created FlowFiles are not transferred to any relationship nor deleted. 

Can we change the file order in the `FetchSlack.testMultipleAttachementsWithFailureNoRetry` test to see how that situation is handled? Currently `files` array contains url for 400, and test_name2.png. What if we change the order as test_name2.png then 400?